model/stats/db.go

```diff
@@ -0,0 +1,572 @@
+	_, database, err := db.GetGlobalSessionFactory().GetSession()
+	if err != nil {
+		err = errors.Wrap(err, "Error establishing db connection")
+		return nil, err
+	}
+	collection := database.C(collectionName)
+	bulk := collection.Bulk()
+	bulk.Unordered()
+	writer := documentWriter{
+		collection:    collection,
+		bulk:          bulk,
+		numOperations: 0,
+	}
+	return &writer, nil
+}
+
+// Adds a document to write to the queue and flush if necessary.
+func (d *documentWriter) write(doc bson.RawD) error {
+	d.bulk.Upsert(bson.M{"_id": getDocId(doc)}, doc)
+	d.numOperations += 1
+
+	if d.numOperations >= bulkSize {
+		err := d.flush()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Flushes all pending operations.
+func (d *documentWriter) flush() error {
+	if d.numOperations > 0 {
+		_, err := d.bulk.Run()
+		if err != nil {
+			return errors.Wrap(err, "Failed to run bulk write")
+		}
+		d.numOperations = 0
+		d.bulk = d.collection.Bulk()
+	}
+	return nil
+}
+
+// Flushes all pending operations and closes the document writer.
+// The document writer should not be used after this method is called.
+func (d *documentWriter) close() error {
+	defer d.collection.Database.Session.Close()
+	err := d.flush()
+	if err != nil {
+		return errors.Wrap(err, "Failed to flush writes")
+	}
+	return nil
+}
+
+// Runs an aggregation pipeline on a collection and calls the provided callback for each output document.
+func aggregateWithCallback(collection string, pipeline []bson.M, callback func(bson.RawD) error) error {
+	session, database, err := db.GetGlobalSessionFactory().GetSession()
+	if err != nil {
+		err = errors.Wrap(err, "Error establishing db connection")
+		return err
+	}
+	defer session.Close()
+
+	session.SetSocketTimeout(0)
+	pipe := database.C(collection).Pipe(pipeline).AllowDiskUse()
+	iter := pipe.Iter()
+	for {
+		raw := bson.RawD{}
+		if iter.Next(&raw) {
+			err = callback(raw)
+			if err != nil {
+				return errors.Wrap(err, "A callback call failed")
+			}
+		} else {
+			err = iter.Err()
+			if err != nil {
+				err = errors.Wrap(err, "Error during aggregation")
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+// Runs an aggregation pipeline on a collection and bulk upserts all the documents into the target collection.
+func aggregateIntoCollection(collection string, pipeline []bson.M, outputCollection string) error {
+	writer, err := newDocumentWriter(outputCollection)
+	if err != nil {
+		return errors.Wrap(err, "Failed to initialize document writer")
+	}
+	err = aggregateWithCallback(collection, pipeline, writer.write)
+	if err != nil {
+		return errors.Wrap(err, "Failed to aggregate with document writer callback")
+	}
+	err = writer.close()
+	if err != nil {
+		return errors.Wrap(err, "Failed to close document writer")
+	}
+	return nil
+}
+
+// Internal function that extracts the _id field from a raw bson document.
+func getDocId(rawDoc bson.RawD) interface{} {
```

Comment:
We cannot limit it to strings, the ids are documents.

Reply:
I realize it's a possibility, but practically speaking these will always be objectIDs or strings. So it would be better to have 2 functions, 1 for when you're working with collections that have string IDs and 1 for objectIDs

Reply:
What I meant is that in this package the ids are always documents, and this is an internal function.

But this function has been incorporated into anser by @tychoish, as part of a new BufferedWriter implementation that I'll use here. So I'll remove this code and the DocumentWriter in a following commit.